### DISCUSSION TITLE
Debugging stuck kernels

### DISCUSSION THREAD

**Discussion by miohtama** [sagemathinc/cocalc#6733] [Q&A] [answered]
What would be the best way to debug a misbehaving kernel (ipykernel).

Sometimes Kernel dies and sometimes it does not start and is stuck in the starting phase.

![image](https://github.com/sagemathinc/cocalc/assets/49922/e90f337e-be74-4071-bacc-57b909e3b160)

I did not find any user interface-facing way to diagnose what could be wrong with the kernel. I have also not seen any diagnosing/debugging options with any other Jupyter implementations. 

Running `cocalc v2` Docker image.

- Can I view the logs / map them to the host filesystem
- Would it be possible to get logs to Docker stdout and use Docker's own logging system for tracking this (seems to be a common practice)

**Comment by haraldschilly** [comment]
I usually try in a terminal: `jupyter console --kernel=[name]`, where the "name" comes from the list given by `jupyter kernelspec list`.

**Reply by williamstein** [reply, marked as answer]
Yes, that's exactly what I do first as well.  It's not 100%, but it helps with any generic problems that would impact all Jupyter clients.

If it's something fairly cocalc specific, enabling this DEBUG flag can be helpful:

https://github.com/sagemathinc/cocalc/blob/master/src/packages/jupyter/kernel/kernel.ts#L27

Note that I just massively refactored the Jupyter code in cocalc during the last week, so if you don't build a new image you might not see packages/jupyter.  This is in master now though.
